src/backend/codegen/include/codegen/base_codegen.h

```diff
@@ -0,0 +1,158 @@
+//---------------------------------------------------------------------------
+//  Greenplum Database
+//  Copyright (C) 2016 Pivotal Software, Inc.
+//
+//  @filename:
+//    base_codegen.h
+//
+//  @doc:
+//    Base class for all the code generators with common implementation
+//
+//---------------------------------------------------------------------------
+#ifndef GPCODEGEN_BASE_CODEGEN_H_  // NOLINT(build/header_guard)
+#define GPCODEGEN_BASE_CODEGEN_H_
+
+#include <string>
+#include <vector>
+
+#include "codegen/utils/codegen_utils.h"
+#include "codegen/codegen_interface.h"
+
+namespace gpcodegen {
+
+/** \addtogroup gpcodegen
+ *  @{
+ */
+
+/**
+ * @brief Base code generator with common implementation that other
+ * 		  code generators can use.
+ *
+ * @tparam FuncPtrType Function type for regular / generated functions.
+ **/
+template<class FuncPtrType>
+class BaseCodeGen: public CodeGenInterface {
+ public:
+  /**
+   * @brief Destroys the code generator and reverts back to using regular
+   * 		version of the generated function.
+   **/
+  virtual ~BaseCodeGen() {
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+  }
+
+  bool GenerateCode(gpcodegen::CodeGenUtils* codegen_utils) final {
```

Comment:
This is overriding virtual function in interface. We have necessary comments in the base class